File: web/src/lib/stores/watchlist.ts
import { writable, derived, get } from 'svelte/store';
import { browser } from '$app/environment';
import { serverCall } from '$lib/api/server';
import type {
	WatchlistEntry,
	WatchlistEntryWithAircraft,
	Aircraft,
	DataListResponse,
	DataResponse
} from '$lib/types';
import { FixFeed } from '$lib/services/FixFeed';

interface WatchlistState {
	entries: WatchlistEntryWithAircraft[];
	loading: boolean;
	error: string | null;
}

const initialState: WatchlistState = {
	entries: [],
	loading: false,
	error: null
};

const watchlistStore = writable<WatchlistState>(initialState);

export const watchlist = {
	subscribe: watchlistStore.subscribe,

	/**
	 * Load watchlist from API
	 */
	async load(): Promise<void> {
		if (!browser) return;

		watchlistStore.update((state) => ({ ...state, loading: true, error: null }));

		try {
			const response = await serverCall<DataListResponse<WatchlistEntry>>('/watchlist', {
				method: 'GET'
			});
			const entries = response.data;

			// Fetch aircraft details for each entry, removing invalid entries
			const entriesWithAircraft: WatchlistEntryWithAircraft[] = [];
			const entriesToRemove: string[] = [];

			await Promise.all(
				entries.map(async (entry) => {
					try {
						const aircraftResponse = await serverCall<DataResponse<Aircraft>>(
							`/aircraft/${entry.aircraftId}`,
							{
								method: 'GET'
							}
						);
						entriesWithAircraft.push({ ...entry, aircraft: aircraftResponse.data });
					} catch {
						// Aircraft not found or invalid - mark for removal
						entriesToRemove.push(entry.aircraftId);
					}
				})
			);

			// Remove invalid entries from the database
			await Promise.all(
				entriesToRemove.map(async (aircraftId) => {
					try {
						await serverCall(`/watchlist/${aircraftId}`, {
							method: 'DELETE'
						});
					} catch {
						// Ignore errors when cleaning up - entry may already be gone
					}
				})
			);

			watchlistStore.set({
				entries: entriesWithAircraft,
				loading: false,
				error: null
			});

			// Update WebSocket subscriptions
			notifyWatchlistChange(entriesWithAircraft);
		} catch (err) {
			const message = err instanceof Error ? err.message : 'Failed to load watchlist';
			watchlistStore.update((state) => ({
				...state,
				loading: false,
				error: message
			}));
		}
	},

	/**
	 * Add aircraft to watchlist
	 */
	async add(aircraftId: string, sendEmail: boolean = false): Promise<void> {
		if (!browser) return;

		try {
			await serverCall<DataResponse<WatchlistEntry>>('/watchlist', {
				method: 'POST',
				body: JSON.stringify({ aircraftId: aircraftId, sendEmail: sendEmail })
			});

			// Reload watchlist
			await this.load();
		} catch (err) {
			const message = err instanceof Error ? err.message : 'Failed to add to watchlist';
			watchlistStore.update((state) => ({ ...state, error: message }));
			throw err;
		}
	},

	/**
	 * Remove aircraft from watchlist
	 */
	async remove(aircraftId: string): Promise<void> {
		if (!browser) return;

		try {
			await serverCall(`/watchlist/${aircraftId}`, {
				method: 'DELETE'
			});

			// Remove from local state immediately
			watchlistStore.update((state) => {
				const entries = state.entries.filter((e) => e.aircraftId !== aircraftId);
				notifyWatchlistChange(entries);
				return { ...state, entries };
			});
		} catch (err) {
			const message = err instanceof Error ? err.message : 'Failed to remove from watchlist';
			watchlistStore.update((state) => ({ ...state, error: message }));
			throw err;
		}
	},

	/**
	 * Update email preference for aircraft
	 */
	async updateEmailPreference(aircraftId: string, sendEmail: boolean): Promise<void> {
		if (!browser) return;

		try {
			await serverCall(`/watchlist/${aircraftId}`, {
				method: 'PUT',
				body: JSON.stringify({ sendEmail: sendEmail })
			});

			// Update local state
			watchlistStore.update((state) => {
				const entries = state.entries.map((e) =>
					e.aircraftId === aircraftId ? { ...e, sendEmail: sendEmail } : e
				);
				return { ...state, entries };
			});
		} catch (err) {
			const message = err instanceof Error ? err.message : 'Failed to update email preference';
			watchlistStore.update((state) => ({ ...state, error: message }));
			throw err;
		}
	},

	/**
	 * Clear entire watchlist
	 */
	async clear(): Promise<void> {
		if (!browser) return;

		try {
			await serverCall('/watchlist/clear', {
				method: 'DELETE'
			});

			watchlistStore.set({
				entries: [],
				loading: false,
				error: null
			});

			notifyWatchlistChange([]);
		} catch (err) {
			const message = err instanceof Error ? err.message : 'Failed to clear watchlist';
			watchlistStore.update((state) => ({ ...state, error: message }));
			throw err;
		}
	},

	/**
	 * Check if aircraft is in watchlist
	 */
	has(aircraftId: string): boolean {
		const state = get(watchlistStore);
		return state.entries.some((e) => e.aircraftId === aircraftId);
	},

	/**
	 * Get active aircraft IDs for WebSocket subscriptions
	 */
	getActiveAircraftIds(): string[] {
		const state = get(watchlistStore);
		return state.entries.map((e) => e.aircraftId);
	}
};

/**
 * Notify FixFeed about watchlist changes for WebSocket subscriptions
 */
function notifyWatchlistChange(entries: WatchlistEntryWithAircraft[]) {
	const aircraftIds = entries.map((e) => e.aircraftId);
	const fixFeedInstance = FixFeed.getInstance();
	fixFeedInstance.subscribeToWatchlist(aircraftIds);
}

/**
 * Derived store for active watchlist entries (all entries are active now)
 */
export const activeWatchlist = derived(watchlistStore, ($store) => $store.entries);

File: src/actions/watchlist.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Json},
};
use tracing::error;
use uuid::Uuid;

use crate::actions::{DataListResponse, DataResponse};
use crate::auth::AuthUser;
use crate::watchlist::{AddToWatchlistRequest, UpdateWatchlistRequest};
use crate::watchlist_repo::WatchlistRepository;
use crate::web::AppState;

/// GET /data/watchlist - Get current user's watchlist
pub async fn get_watchlist(
    auth_user: AuthUser,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let repo = WatchlistRepository::new(state.pool);
    let user_id = auth_user.0.id;

    match repo.get_by_user(user_id).await {
        Ok(entries) => Json(DataListResponse { data: entries }).into_response(),
        Err(e) => {
            error!("Failed to get watchlist: {}", e);
            (StatusCode::INTERNAL_SERVER_ERROR, "Failed to get watchlist").into_response()
        }
    }
}

/// POST /data/watchlist - Add aircraft to watchlist
pub async fn add_to_watchlist(
    auth_user: AuthUser,
    State(state): State<AppState>,
    Json(req): Json<AddToWatchlistRequest>,
) -> impl IntoResponse {
    let repo = WatchlistRepository::new(state.pool);
    let user_id = auth_user.0.id;

    match repo.add(user_id, req.aircraft_id, req.send_email).await {
        Ok(entry) => (StatusCode::CREATED, Json(DataResponse { data: entry })).into_response(),
        Err(e) => {
            error!("Failed to add to watchlist: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to add to watchlist",
            )
                .into_response()
        }
    }
}

/// PUT /data/watchlist/{aircraft_id} - Update email preference
pub async fn update_watchlist_email(
    auth_user: AuthUser,
    State(state): State<AppState>,
    Path(aircraft_id): Path<Uuid>,
    Json(req): Json<UpdateWatchlistRequest>,
) -> impl IntoResponse {
    let repo = WatchlistRepository::new(state.pool);
    let user_id = auth_user.0.id;

    match repo
        .update_email_preference(user_id, aircraft_id, req.send_email)
        .await
    {
        Ok(true) => StatusCode::NO_CONTENT.into_response(),
        Ok(false) => (StatusCode::NOT_FOUND, "Watchlist entry not found").into_response(),
        Err(e) => {
            error!("Failed to update watchlist: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to update watchlist",
            )
                .into_response()
        }
    }
}

/// DELETE /data/watchlist/{aircraft_id} - Remove from watchlist
pub async fn remove_from_watchlist(
    auth_user: AuthUser,
    State(state): State<AppState>,
    Path(aircraft_id): Path<Uuid>,
) -> impl IntoResponse {
    let repo = WatchlistRepository::new(state.pool);
    let user_id = auth_user.0.id;

    match repo.remove(user_id, aircraft_id).await {
        Ok(true) => StatusCode::NO_CONTENT.into_response(),
        Ok(false) => (StatusCode::NOT_FOUND, "Watchlist entry not found").into_response(),
        Err(e) => {
            error!("Failed to remove from watchlist: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to remove from watchlist",
            )
                .into_response()
        }
    }
}

/// DELETE /data/watchlist - Clear entire watchlist
pub async fn clear_watchlist(
    auth_user: AuthUser,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let repo = WatchlistRepository::new(state.pool);
    let user_id = auth_user.0.id;

    match repo.clear_user(user_id).await {
        Ok(_count) => StatusCode::NO_CONTENT.into_response(),
        Err(e) => {
            error!("Failed to clear watchlist: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to clear watchlist",
            )
                .into_response()
        }
    }
}
